### PR TITLE
OpenSora-pku v1.1: improve training speed and documents

### PR DESCRIPTION
**Pull request by wtomin**
To accelerate training speed, major changes are made:
  1. Update CANN version ([CANN C18(0517)](https://repo.mindspore.cn/ascend/ascend910/20240517/)->[CANN C18(0705)](https://repo.mindspore.cn/ascend/ascend910/20240705/)) and MindSpore version ([MS2.3_master(0615)](https://repo.mindspore.cn/mindspore/mindspore/version/202406/20240615/master_20240615020018_43ccb91e45899b64fe31d304497ab17e3ada3cea_newest/unified/) -> [MS2.3_master(0705)](https://repo.mindspore.cn/mindspore/mindspore/version/202407/20240705/master_20240705220018_51f414917fd9a312dd43ea62eea61cf37c3dfbd6_newest/unified/)); inference results on the new versions are visually fine. The overfitting experiment generates reasonable results.
  2. Enable optimizer parallelism fusion by default (See `enable_parallel_fusion` in train args);
  3. Change FlashAttention default layout from `BNSD` to `BSH` (See `mindone.models.modules.flash_attention`);
  4. Set `num_no_recompute` to $4$ in `scripts/text_condition/train_videoae_65x512x512_img16.sh`.

Minor changes are also made to:
1. `train_t2v.py`: added some arguments assertions to ensure the input arguments are valid;
2. `t2v_dataset.py`: change the mask dtype from uint8 to int32, because cann 0705 does not support tiling with uint8 tensors.
3. `README.md`: include two more documents (performance boosting history and training args) to enhance users' understanding; include a single-device training script `train_videoae_65x512x512_single_device.sh` as an example.

### CODE REVIEW

**Review by SamitHuang — APPROVED** [review]

File: examples/opensora_pku/docs/performance_boosting_history.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+Performance Boosting History
+
+In this doc, we record the performance history of Open-Sora-Plan v1.1.0.
+History:
+
+
+
+- 2024.07.16
+  Major Updates:
```
Comment (CaitinZhao) [inline examples/opensora_pku/docs/performance_boosting_history.md:9]: 是不是附上升级的PR

File: examples/opensora_pku/opensora/train/train_t2v.py [hunk 2]
```diff
@@ -437,7 +523,7 @@ def main(args):
 
     # 6. train
     model.train(
-        args.epochs,
+        sink_epochs,
```
Comment (CaitinZhao) [inline examples/opensora_pku/opensora/train/train_t2v.py:526]: 如果改这个的话，callback里的打印也要改一下

File: examples/opensora_pku/docs/performance_boosting_history.md (hunk 1 above)
Comment (wtomin) [reply examples/opensora_pku/docs/performance_boosting_history.md]: 已修改

File: examples/opensora_pku/opensora/train/train_t2v.py (hunk 2 above)
Comment (wtomin) [reply examples/opensora_pku/opensora/train/train_t2v.py:526]: EvalSaveCallback里面的`cb_params.batch_num` 和`cb_params.epoch_num`似乎对应的是`sink_size`和`sink_epochs`. 这段代码里：
```python
class EvalSaveCallback(Callback)
    def on_train_step_end(self, run_context):
        cb_params = run_context.original_args()
        loss = _handle_loss(cb_params.net_outputs)
        cur_step = cb_params.cur_step_num + self.start_epoch * cb_params.batch_num
        step_num = cb_params.batch_num * cb_params.epoch_num

        if cur_step % cb_params.batch_num == 0:
            cur_epoch = cb_params.cur_epoch_num
        else:
            cur_epoch = cb_params.cur_epoch_num - 1

```

涉及到epoch id或者step id的都需要根据`sink_size` 和 `dataset_size`的关系换算一下。是这意思不？

File: examples/opensora_pku/opensora/train/train_t2v.py [hunk 3]
```diff
@@ -248,32 +263,100 @@ def main(args):
...
+        total_train_steps = args.max_train_steps
+        args.epochs = total_train_steps // dataset_size
+    else:
+        # use args.epochs
+        assert (
+            args.epochs is not None and args.epochs > 0
+        ), f"When args.max_train_steps is not provided, args.epochs must be a positive integer! but got {args.epochs}"
+        total_train_steps = args.epochs * dataset_size
+
+    sink_epochs = total_train_steps // steps_per_sink
```
Comment (SamitHuang) [inline examples/opensora_pku/opensora/train/train_t2v.py:293]: sink_epochs = math.ceil(total_train_steps / steps_per_sink)

File: examples/opensora_pku/opensora/train/train_t2v.py [hunk 4]
```diff
@@ -248,32 +263,100 @@ def main(args):
...
+    else:
+        steps_per_sink = dataset_size
+
+    if args.max_train_steps is not None:
+        assert args.max_train_steps > 0, f"max_train_steps should a positive integer, but got {args.max_train_steps}"
+        assert (
+            args.max_train_steps >= steps_per_sink
+        ), f"Expect that the max_train_steps is no less than {steps_per_sink}, but got {args.max_train_steps}"
+        total_train_steps = args.max_train_steps
+        args.epochs = total_train_steps // dataset_size
```
Comment (SamitHuang) [inline examples/opensora_pku/opensora/train/train_t2v.py:285]: args.epochs = math.ceil(total_train_steps / steps_per_sink)

should allow set max_train_steps if only it's >0

File: examples/opensora_pku/opensora/train/train_t2v.py (hunk 3 above)
Comment (wtomin) [reply examples/opensora_pku/opensora/train/train_t2v.py:293]: `max_train_steps` is the **maximum** training steps, meaning `total_train_steps <=max_train_steps`. Therefore the `sink_epochs` should be the floor not the ceiling of `max_train_steps/step_per_sink`. 

To avoid the `sink_epochs` being zero, there is an assertion:
```python
        assert (
            args.max_train_steps >= steps_per_sink
        ), f"Expect that the max_train_steps is no less than {steps_per_sink}, but got {args.max_train_steps}"
        total_train_steps = args.max_train_steps
        args.epochs = total_train_steps // dataset_size
```
If the user gives too small `args.max_train_steps`, the program will raise an error, and ask the user to increase `args.max_train_steps` to the value of `steps_per_sink`.
I will further improve the printing message.

Comment (wtomin) [reply examples/opensora_pku/opensora/train/train_t2v.py]: Fixed it according to your suggestion.

File: examples/opensora_pku/opensora/train/train_t2v.py (hunk 4 above)
Comment (wtomin) [reply examples/opensora_pku/opensora/train/train_t2v.py]: Fixed it according to your suggestion.

File: examples/opensora_pku/opensora/train/train_t2v.py (hunk 2 above)
Comment (wtomin) [reply examples/opensora_pku/opensora/train/train_t2v.py:518]: 理论上， cb_params.cur_step_num ， self.start_epoch，cb_params.batch_num这三个应该是sink_epochs，sink_size配置后新算的。 所以print 出来的epoch id和 step id也是 based on sink_epochs和 sink_size。

但是我代码打印有点问题，epoch id 和 step id， loss这些不会以一定间隔打印在屏幕上， 所以用户感知不到。

最后loss会记录在`result.log`里面：
```
step	loss	train_time(s)
31	0.0013550	769.2326651
62	0.1211518	76.9347382
93	0.0235147	76.6365781
124	0.0529454	76.7953396
```
这里只有整个训练途中的step id，不需要换算了。
况且这个epoch id和 step id 错配的问题，只在sink_size!=dataset_size的情况下出现。 目前代码默认的sink_size都是等于dataset_size的， 这个问题暂时优先级不高。

**Review by CaitinZhao — APPROVED** [review]
